src/librustc_mir/dataflow/impls/requires_storage.rs

```diff
@@ -0,0 +1,112 @@
+pub use super::*;
+
+use rustc::mir::*;
+use crate::dataflow::BitDenotation;
+
+/// A conservative approximation of whether each local requires storage at each
+/// point.
+///
+/// This is used to limit which locals are kept in the generator state.
+///
+/// Currently a local is consider to possibly require storage if:
+/// * Any part of it has ever been initialized - borrowck should ensure that
+///   this ends up the same as it's ever been used. If we add a way to
+///   reference uninitialized variables then that would also `gen` this.
+///   This would also be a problem if we ran the generator transform after
+///   deaggregation.
+/// * And if it hasn't had it's storage killed, or been droped.
+///
+/// Notably we don't assume that moving from a variable is not considered here.
+/// It's possible that something like the following could happen:
+///
+/// let mut x = String::new();
+/// let p = &mut x as *mut String;
+/// let y = x;
+/// p.write(String::new());             // It's not decided whether this is UB
+#[derive(Copy, Clone)]
+pub struct MaybeRequiresStorage<'a, 'tcx: 'a> {
+    mir: &'a Body<'tcx>,
+}
+
+impl<'a, 'tcx: 'a> MaybeRequiresStorage<'a, 'tcx> {
+    pub fn new(mir: &'a Body<'tcx>)
+               -> Self {
+        MaybeRequiresStorage { mir }
+    }
+
+    pub fn mir(&self) -> &Body<'tcx> {
+        self.mir
+    }
+}
+
+impl<'a, 'tcx> BitDenotation<'tcx> for MaybeRequiresStorage<'a, 'tcx> {
+    type Idx = Local;
+    fn name() -> &'static str { "maybe_storage_live" }
+    fn bits_per_block(&self) -> usize {
+        self.mir.local_decls.len()
+    }
+
+    fn start_block_effect(&self, _sets: &mut BitSet<Local>) {
+        // Nothing is live on function entry
+    }
+
+    fn statement_effect(&self,
+                        sets: &mut BlockSets<'_, Local>,
+                        loc: Location) {
+        let stmt = &self.mir[loc.block].statements[loc.statement_index];
+
+        match stmt.kind {
+            StatementKind::Assign(ref place, _)
+            | StatementKind::SetDiscriminant { ref place, .. } => {
+                place.base_local().map(|l| sets.gen(l));
+            }
+            StatementKind::InlineAsm(box InlineAsm { ref outputs, .. }) => {
+                for p in &**outputs {
+                    p.base_local().map(|l| sets.gen(l));
+                }
+            }
+            StatementKind::StorageDead(l) => sets.kill(l),
+            _ => (),
+        }
+    }
+
+    fn terminator_effect(&self,
+                         sets: &mut BlockSets<'_, Local>,
+                         loc: Location) {
+        match &self.mir[loc.block].terminator().kind {
+            TerminatorKind::DropAndReplace { location: _, .. } => {
+                debug_assert!(false, "Generator transform should be run after drop elaboration");
```

Comment:
Let's not ever do this in the compiler, if the condition is simple enough we should use `span_bug!`.